extensions/azdata/src/azdata.ts

```diff
@@ -157,22 +186,68 @@ export class AzdataTool implements azdataExt.IAzdataApi {
 					if (args.memoryRequest) { argsArray.push('--memory-request', args.memoryRequest); }
 					if (args.noWait) { argsArray.push('--no-wait'); }
-					return this.executeCommand<void>(argsArray, additionalEnvVars);
+					return this.executeCommand<void>(argsArray, additionalEnvVars, session);
 				}
 			}
 		}
 	};
 
-	public login(endpoint: string, username: string, password: string, additionalEnvVars: azdataExt.AdditionalEnvVars = {}): Promise<azdataExt.AzdataOutput<void>> {
-		return this.executeCommand<void>(['login', '-e', endpoint, '-u', username], Object.assign({}, additionalEnvVars, { 'AZDATA_PASSWORD': password }));
+	public async login(endpoint: string, username: string, password: string, additionalEnvVars: azdataExt.AdditionalEnvVars = {}): Promise<azdataExt.AzdataOutput<void>> {
+		// Since login changes the context we want to wait until all currently executing commands are finished before this is executed
+		while (this._currentlyExecutingCommands.length > 0) {
+			await this._currentlyExecutingCommands[0];
+		}
+		// Logins need to be done outside the session aware logic so call impl directly
+		return this.executeCommandImpl<void>(['login', '-e', endpoint, '-u', username], Object.assign({}, additionalEnvVars, { 'AZDATA_PASSWORD': password }));
+	}
+
+	public async acquireSession(endpoint: string, username: string, password: string, additionalEnvVars?: azdataExt.AdditionalEnvVars): Promise<azdataExt.AzdataSession> {
+		const session = new AzdataSession();
+		session.sessionEnded().then(async () => {
+			// Wait for all commands running for this session to end
+			while (this._currentlyExecutingCommands.length > 0) {
+				await this._currentlyExecutingCommands[0].promise;
+			}
+			this._currentSession = undefined;
+			// Start our next command now that we're all done with this session
+			// TODO: Should we check if the command has a session that hasn't started? That should never happen..
+			// TODO: Look into kicking off multiple commands
+			this._queuedCommands.shift()?.deferred.resolve();
+		});
+
+		// We're not in a session or waiting on anything so just set the current session right now
+		if (!this._currentSession && this._queuedCommands.length === 0) {
+			this._currentSession = session;
+		} else {
+			// We're in a session or another command is executing so add this to the end of the queued commands and wait our turn
+			const deferred = new Deferred<void>();
+			deferred.promise.then(() => {
+				this._currentSession = session;
+				// We've started a new session so look at all our queued commands and start
+				// the ones for this session now.
+				this._queuedCommands = this._queuedCommands.filter(c => {
+					if (c.session === this._currentSession) {
+						c.deferred.resolve();
+						return false;
+					}
+					return true;
+				});
+			});
+			this._queuedCommands.push({ deferred, session: undefined });
+			await deferred.promise;
+		}
+
+		await this.login(endpoint, username, password, additionalEnvVars);
```

Comment:
Does this PR reduce the number of times we azdata login?  Azdata login used to be executed before each azdata call, which is unnecessary.  The logins are good for 10 hours (you can see the expiration in ~/.azdata/azdata-profile.json).  We could speed things up by either checking the expiration, or only logging in when we catch an http 401.  Preferable the later.

Reply:
No, and we can't easily reduce that since we can't know for sure that the current azdata login context is the "correct" one unless we login for every set of commands we execute (this is much more of an issue now that we support multiple controllers - although someone doing an azdata login outside of ADS would also cause this problem)

The end solution that's being worked on is to be able to set the context for a specific command through parameters passed to azdata - which would then apply to that session only. But until that work is done we need to take the perf hit and do this or we risk executing commands against the wrong cluster.